src/coreclr/jit/lsraarm64.cpp

```diff
@@ -25,6 +25,262 @@ XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
+    }
+
+    assert(refPosCount == firstRefPosition->regCount);
+
+    return true;
+}
+
+//------------------------------------------------------------------------
+// areNextConsecutiveRegistersFree: Starting with `regToAssign`, check if next
+//   consecutive `registersToCheck` are free or not.
+//
+// Arguments:
+//    regToAssign     - Register assigned to the first refposition.
+//    registersCount  - Number of registers to check.
+//    registerType    - Type of register.
+//
+//  Returns:
+//      True if all the consecutive registers starting from `regToAssign` were free. Even if one
+//      of them is busy, returns false.
+//
+bool LinearScan::areNextConsecutiveRegistersFree(regNumber regToAssign, int registersCount, var_types registerType)
+{
+    for (int i = 0; i < registersCount; i++)
+    {
+        if (isRegInUse(regToAssign, registerType))
+        {
+            return false;
+        }
+        regToAssign = regToAssign == REG_FP_LAST ? REG_FP_FIRST : REG_NEXT(regToAssign);
+    }
+
+    return true;
+}
+
+//------------------------------------------------------------------------
+// getFreeCandidates: Returns the mask of all the free candidates for given refPosition.
+//   If refPosition is the first RefPosition of a series of refpositions that needs
+//   consecutive registers, then returns only the mask such that it satisfies the need
+//   of having free consecutive registers.
+//
+// Arguments:
+//    candidates   - Register assigned to the first refposition.
+//    refPosition  - Number of registers to check.
+//
+//  Returns:
+//      Register mask of all the free registers
+//
+regMaskTP LinearScan::getFreeCandidates(regMaskTP candidates, RefPosition* refPosition)
+{
+    regMaskTP result = candidates & m_AvailableRegs;
+    if (!refPosition->isFirstRefPositionOfConsecutiveRegisters())
+    {
+        return result;
+    }
+
+    unsigned int registersNeeded   = refPosition->regCount;
+    regMaskTP    currAvailableRegs = result;
+    if (BitOperations::PopCount(currAvailableRegs) < registersNeeded)
+    {
+        // If number of free registers are less than what we need, no point in scanning
+        // for them.
+        return RBM_NONE;
+    }
+
+// At this point, for 'n' registers requirement, if Rm+1, Rm+2, Rm+3, ..., Rm+k are
+// available, create the mask only for Rm+1, Rm+2, ..., Rm+(k-n+1) to convey that it
+// is safe to assign any of those registers, but not beyond that.
+#define AppendConsecutiveMask(startIndex, endIndex, availableRegistersMask)                                            \
+    regMaskTP selectionStartMask = (1ULL << regAvailableStartIndex) - 1;                                               \
+    regMaskTP selectionEndMask   = (1ULL << (regAvailableEndIndex - registersNeeded + 1)) - 1;                         \
+    consecutiveResult |= availableRegistersMask & (selectionEndMask & ~selectionStartMask);                            \
```

Comment:
Isn't it the case here that `availableRegistersMask` *must* have all the bits set that `(selectionEndMask & ~selectionStartMask)` has set?

Reply:
That's right, and we want to just extract the 1st of those bits to be set in `consecutiveResult `